Keeper retries during insert (clean)

devcrafter opened the pull request:
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Support for keeper request retries during insert into replicated merge trees. Apart from fault tolerance, it aims to provide better user experience, - avoid returning a user an error during insert if keeper is restarted (for example, due to upgrade).

It also introduces failure injection mode for keeper requests during insert into replicated merge tree, so it's possible to inject a failure with certain probability when accessing keeper interface (see `insert_keeper_fault_injection_probability`). It's also possible to reproduce a particular insert execution with failure injections by providing the same seed, see `insert_keeper_fault_injection_seed`

Closes https://github.com/ClickHouse/ClickHouse/issues/39764

It should be merged instead of #41734. This PR contains all the changes from #41734, but commit history does less harm to git blame output (some details in comment to [5401300](https://github.com/ClickHouse/ClickHouse/pull/42607/commits/54013009d3649051eacf21debae7093ae4968621))

devcrafter committed:
Code migration from #41734 to align diff and git blame.

Initial implementation was different and it changed the entire
ReplicatedMergeTreeSink::commitPart() which change history provided by git blame.
Then RetriesControl.retryLoop() was introduced later which significantly reduces
the diff since it's like while() used before.
So, check outing the current version will keep more original history in
git blame, which is useful here

devcrafter committed:
Disable fault injections by default

+ enable fault injection for functional tests

devcrafter commented:
#42646 created regarding AST fuzzer finding

devcrafter commented on src/Storages/MergeTree/ZooKeeperWithFaultInjection.h:
```diff
@@ -0,0 +1,429 @@
+///
+/// ZooKeeperWithFailtInjection mimics ZooKeeper interface and inject failures according to failure policy if set
+///
+class ZooKeeperWithFaultInjection
```
I didn't polish it, but @antonio2368 helped by providing a patch. Now it looks more concise. It's on old branch. I'll move everything to this branch soon

devcrafter commented on src/Storages/MergeTree/ZooKeeperWithFaultInjection.h:
```diff
@@ -0,0 +1,429 @@
+                    fault_policy->beforeOperation();
+            }
+
+            if constexpr (!std::is_same_v<Result, void>)
```
It'd be definitely be nice w/o this `if` but unfortunately `void` is not regular type in C++. So, it's forced by language. Regarding rest if-statements I'd keep it for now, later we can try to make it better readable.

tavplubix replied:
> later we can try to make it better readable

Ha-ha, the truth is that probably we will never rewrite it once it merged, so it's kind of self-deception :)
But I don't really care, because this code is isolated, so it's okay to just keep it.

devcrafter replied:
There is no truth here, only probability :)

Actually, I don't really see right now how can readability be improved significantly here. @alesapin wanted to use this fault injection in other places, and since the current implementation doesn't cover the entire ZK client interface, there is a chance we'd need to cover more functions. It can be a point where some improvements can be done here

P.S. I was able to delete `else` statement. Feels good

devcrafter commented on src/Storages/MergeTree/ZooKeeperWithFaultInjection.h:
```diff
@@ -0,0 +1,429 @@
+
+    template <
+        typename Result,
+        bool no_throw_access = false,
```
We can, but we know it on compile time, so that's why I put it as template argument. It's not used often

devcrafter commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -535,18 +613,26 @@ void ReplicatedMergeTreeSink::commitPart(
+             */
             transaction.commit();
-            storage.enqueuePartForCheck(part->name, MAX_AGE_OF_LOCAL_PART_THAT_WASNT_ADDED_TO_ZOOKEEPER);
+            part_committed_locally_but_zookeeper = true;
```
It's still valid English in my knowledge

devcrafter commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -535,18 +613,26 @@ void ReplicatedMergeTreeSink::commitPart(
+                ErrorCodes::UNKNOWN_STATUS_OF_INSERT,
+                "Unknown status, client must retry. Reason: {}",
+                Coordination::errorMessage(multi_code));
+            return;
```
Yes, it's return from lamba provided to `retryLoop()`

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

devcrafter committed:
Sync with retiring branch igor/insert_zk_retries

devcrafter committed:
Fix flaky test

devcrafter committed:
Add clean up for faults injected before zk request

+ for multi-op request, it create responses in multi response
+ some polishing

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

devcrafter committed:
Fix for messed up test output file

devcrafter commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -334,16 +359,49 @@ void ReplicatedMergeTreeSink::commitPart(
+        /// but operation could be completed by keeper server
+        if (retries_ctl.isRetry())
+        {
+            if (part_committed_locally_but_zookeeper)
```
/// If this flag is true, then part is in Active state, and we'll not retry anymore
/// we only check if part was committed to zk and return success or failure correspondingly
/// Note: if commit to zk failed, then cleanup thread will mark the part as Outdated later

devcrafter committed:
Add more comments

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

devcrafter committed:
Fix style check

tavplubix commented on src/Storages/MergeTree/PartMovesBetweenShardsOrchestrator.cpp:
```diff
@@ -5,6 +5,7 @@
 #include <Poco/JSON/JSON.h>
 #include <Poco/JSON/Object.h>
 #include <Poco/JSON/Parser.h>
+#include <Storages/MergeTree/ZooKeeperWithFaultInjection.h>
```
this include looks redundant

tavplubix commented on src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp:
```diff
@@ -634,12 +737,13 @@ void ReplicatedMergeTreeSink::onStart()
 void ReplicatedMergeTreeSink::onFinish()
 {
     auto zookeeper = storage.getZooKeeper();
+    /// todo: check this place, afaiu, it can be called after Generate apart of Consume
```
It can (see also https://github.com/ClickHouse/ClickHouse/pull/42830)

devcrafter commented on tests/queries/0_stateless/01158_zookeeper_log_long.sql:
```diff
@@ -20,7 +22,7 @@ select 'parts';
        watch_state, path_created, stat_version, stat_cversion, stat_dataLength, stat_numChildren
 from system.zookeeper_log
-where (session_id, xid) in (select session_id, xid from system.zookeeper_log where path='/test/01158/' || currentDatabase() || '/rmt/replicas/1/parts/all_0_0_0')
+where (session_id, xid) in (select session_id, xid from system.zookeeper_log where path='/test/01158/' || currentDatabase() || '/rmt/replicas/1/parts/all_0_0_0' and op_num not in(3))
```
Here I excluded `Exists` query from request result because AFAIS it's done from background thread and there is no guarantee that it'll be called during. Locally, I observe that `Exists` request is part of the test output for release build and absent for debug build.

Stack example: https://pastila.nl/?016c868b/9f840f5cb1dc2400a9bd466c14340882

tavplubix replied:
> it's done from background thread and there is no guarantee that it'll be called during

Yes, you're right. We can add another `system sync replica rmt` before `system flush logs` to provide this guarantee

devcrafter commented on src/Storages/StorageReplicatedMergeTree.cpp:
```diff
@@ -4460,8 +4460,6 @@ void StorageReplicatedMergeTree::assertNotReadonly() const
 SinkToStoragePtr StorageReplicatedMergeTree::write(const ASTPtr & /*query*/, const StorageMetadataPtr & metadata_snapshot, ContextPtr local_context)
 {
     const auto storage_settings_ptr = getSettings();
-    assertNotReadonly();
```
I think you're right. It looks like some leftover.

devcrafter replied:
Well, I forgot, but this change was done on purpose. We allow creating `ReplicatedMergeTreeSink` even for table in read-only mode since the sink can now handle it with retries.

Currently, integration test for retries is failing because of this check

devcrafter committed:
Fix review comments

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

devcrafter committed:
Better handling tables in read-only mode during insert

+ throw immediately if table has no metadata in zk
+ stop retries in case of shutdown
+ check if table is readonly at the begining of every retry

devcrafter committed:
Mark a test as long

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

devcrafter commented:
> I just wanted to let you know I did some basic tests and didn't find any issues. Basically I've run 600 inserts (one every 0.1 seconds) while I applied chaos engineering (and I was the monkey restarting random ZK instances in a cluster of 3) and it worked without flaws.

@Algunenano Thanks. Your testing is very appreciated

I just wanted to mention that the insert retries are doing the best effort job, so, if you're especially "lucky" the insert still can fail, - for example, - when a request to keeper which corresponds to committing a part to the keeper (see [request](https://github.com/ClickHouse/ClickHouse/pull/42607/files#diff-d34dfcef2ee825d0184c61973c0c3c8606d68975870f22c9ddd1501f7ff88e42R619)) is failed and no data was written. In such case, we'll return `Insert failed due to zookeeper error. Please retry. Reason: ...` from [here](https://github.com/ClickHouse/ClickHouse/pull/42607/files#diff-d34dfcef2ee825d0184c61973c0c3c8606d68975870f22c9ddd1501f7ff88e42R402) (note: the links above are part of a file with large diff, so you'd need to unfold the file first to see the corresponding code lines)

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

antonio2368 commented on src/Storages/MergeTree/ZooKeeperWithFaultInjection.h:
```diff
@@ -0,0 +1,527 @@
+        const char * func_name,
+        const std::string & path,
+        Operation operation,
+        FaultCleanupType<Result> fault_after_op_cleanup = {},
+        FaultCleanupType<void> fault_before_op_cleanup = {})
```
```suggestion
        FaultCleanupType<Result> fault_after_op_cleanup = {},
        FaultCleanupType<void> fault_before_op_cleanup = {})
```

Why do you have 2 arguments?
Also, with this approach you don't need FaultCleanupType, first can be directly `std::function<void(Result &)>` and second `std::function<void()>`

devcrafter commented:
@tavplubix Could you look at `00502_custom_partitioning_replicated_zookeeper_long` failure. It's not really clear to me why it became flaky

devcrafter committed:
Disable keeper retries during INSERT by default

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

devcrafter committed:
Added missed insert_keeper_max_retries setting in test config

devcrafter commented:
`01346_alter_enum_partition_key_replicated_zookeeper_long` failure [here](https://s3.amazonaws.com/clickhouse-test-reports/42607/dd598fdd4a1b7fbbe502f586b0b2d4b8025af4d7/stateless_tests_flaky_check__asan_.html) need investigation

devcrafter committed:
Fixes

+ enable retries in integration test
+ typo in fault injection traces

Commit:
Automatic style fix

devcrafter committed:
Merge branch 'master' into igor/insert_zk_retries_retry

devcrafter committed:
Fix for flaky test

00502_custom_partitioning_replicated_zookeeper_long

devcrafter committed:
Disable fault injection and retries function tests settings

alesapin committed:
Trying to fix two flaky tests

alesapin committed:
better tests

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

devcrafter committed:
fix flaky test

+ 01346_alter_enum_partition_key_replicated_zookeeper_long

alesapin committed:
Fix wrong logic

devcrafter committed:
Merge remote-tracking branch 'origin/master' into igor/insert_zk_retries_retry

tavplubix commented:
@devcrafter, you forgot to take a look at Stress Tests failures, some of them are related

devcrafter commented:
> @devcrafter, you forgot to take a look at Stress Tests failures, some of them are related

Stress test failures were caused by [config](https://github.com/ClickHouse/ClickHouse/pull/42607/files#diff-9e5613b80b115f81df382bc15be97c98c6656cc72d924fd90ce884b7010e2ee1) for functional tests, which includes newly introduced settings. The settings are not known by previous ClickHouse versions, so it failed to start.

The settings are reintroduced in #43117

zhyhang commented:
@devcrafter 
version 23.8.5.16
settings: insert_keeper_max_retries=20 (default value)
insert 72,000 rows to a replicate merge tree table, when kill a zookeeper node, the insert finish normally, but observe duplicated rows (about 85,000)？
